dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -53,38 +55,148 @@ public Pattern getParameterPattern() {
+        String qry = "SELECT COLUMNNAME AS COLUMN_NAME, COLUMNDATATYPE AS TYPE_NAME, "
+                + " ROW_NUMBER() OVER() AS ORDINAL_POSITION, " +
+                + DatabaseMetaData.procedureColumnIn + " AS COLUMN_TYPE "
+                + "'TABLE' AS PARENT_OBJECT_TYPE " 
                 + "FROM SYS.SYSCOLUMNS WHERE REFERENCEID = "
-                + "(SELECT TABLEID FROM SYS.SYSTABLES WHERE TABLENAME = ?)";
-        return readIntoParams(tableOrViewName.toUpperCase(), qry);
+                + "(SELECT TABLEID FROM SYS.SYSTABLES WHERE TABLENAME = ?)"
+                + "ORDER BY COLUMNNUMBER";
+        return fetchIntoParams(tableOrViewName.toUpperCase(), qry);
     }
 
-    private Map<String, DbParameterAccessor> readIntoParams(
-            String tableOrViewName, String query) throws SQLException {
+    private Map<String, DbParameterAccessor> readIntoParams(ResultSet rs)
+                                                      throws SQLException {
+        Map<String, DbParameterAccessor> allParams = new HashMap<>();
+        while (rs.next()) {
+            String columnName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+            DbParameterAccessor dbp =
+                createObjectDbParamAccessor(columnName, rs,
+                                           (rs.getString("PARENT_OBJECT_TYPE") == "FUNCTION" ? true : false));
+            allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+        }
+        return allParams;
+    }
+
+    private Map<String, DbParameterAccessor> fetchIntoParams(String objectName, String query)
+                                                                               throws SQLException {
         try (PreparedStatement dc = getConnection().prepareStatement(query)) {
-            dc.setString(1, tableOrViewName);
-
-            ResultSet rs = dc.executeQuery();
-            Map<String, DbParameterAccessor> allParams = new HashMap<String, DbParameterAccessor>();
-            int position = 0;
-            while (rs.next()) {
-                String columnName = rs.getString(1);
-                String dataType = rs.getString(2);
-                DbParameterAccessor dbp = createDbParameterAccessor(
-                        columnName,
-                        Direction.INPUT,
-                        typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            dc.setString(1, objectName);
+            try (ResultSet rs = dc.executeQuery()) {
+                return readIntoParams(rs);
             }
-            rs.close();
-            return allParams;
         }
     }
 
-    public Map<String, DbParameterAccessor> getAllProcedureParameters(
-            String procName) throws SQLException {
-        throw new UnsupportedOperationException();
+    private Direction getParamDirection(int columnType, boolean isFunction) {
```

Comment:
It should be possible the result to be unified on SQL query level without a need to handle results differently for different types of objects (tables, procedures functions).